MetaMask/metamask-extension
# fix: Retry requests from provider after connection is ready, on all builds

## **Description**

Currently, requests from the dapp can hang if: 
1. they are sent before the extension is fully initialized / before the extension is ready to process requests from that specific dapp connection
2. they get swallowed do to an issue in chromium with streams on prerendered pages breaking when they become visible to the user

The inpage provider does have a retry mechanism, but currently that only works on MV3 builds. Additionally, is not a 100% reliable trigger as the underlying `metamask_chainChanged` message that it relies on is tied to MetaMaskController mem state changes, and is not fired off each time a dapp connection is established.

This PR attempts to fix request retries by:
* Making the retry message triggered by `metamask_chainChanged` work on all builds, not just MV3 builds
* Firing `metamask_chainChanged` every time a new dapp connection is established


**Open concern:**
There may be an edge case when a request with a confirmation is sent by the dapp, then received and shown to the user in the wallet, but swallowed by prerender. I believe my proposed changes in this PR would cause the confirmation to get resent even though there may still be a confirmation being shown to the user. I haven't been able to test this properly as debugging the prerender issue in general is difficult. 

This would require a dapp to send a request requiring confirmation pretty much immediately after page load, implying without user interaction.


[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24142?quickstart=1)

## **Related issues**

See: https://github.com/MetaMask/metamask-extension/issues/11069
See: https://github.com/MetaMask/metamask-extension/issues/23329
See: https://github.com/MetaMask/metamask-extension/issues/24025

## **Manual testing steps**

NOTE: these testing steps require that you add an artificial 5s delay to the background connection handling. See the comments below to find the commit where this delay was included in this PR before being removed for merging

### Verify this fixes the `metamask_sendDomainMetadata` issue
1. Restart extension
2. Open any page
3. Quickly open the wallet popup
4. Notice that there is no favicon
5. After 5s have elapsed since the page was opened, the favicon should populate in the wallet popup

### Verify this fixes retrying messages before extension is ready
1. Open any page (restarting extension not required)
2. Quickly open console and type `await window.ethereum.request({method: 'eth_chainId'})`
3. Notice that the response doesn't come immediately
4. After 5s have elapsed since the page was opened, the response should be shown in console

### Verify this fixes retrying messages after broken prerender stream
Not easy to verify this unfortunately, you would need to break the prerender stream while a request was being made


## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
